File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/progress/ExoPlayerFileProgressReader.java
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.progress;

import com.google.android.exoplayer2.ExoPlayer;
import com.lasthopesoftware.bluewater.client.playback.file.progress.FileProgress;
import com.lasthopesoftware.bluewater.client.playback.file.progress.ReadFileProgress;

public class ExoPlayerFileProgressReader implements ReadFileProgress {

	private final ExoPlayer exoPlayer;

	private FileProgress fileProgress = new FileProgress(0, 0);

	public ExoPlayerFileProgressReader(ExoPlayer exoPlayer) {
		this.exoPlayer = exoPlayer;
	}

	@Override
	public synchronized FileProgress getFileProgress() {
		if (!exoPlayer.getPlayWhenReady()) return fileProgress;

		return fileProgress = new FileProgress(
			exoPlayer.getCurrentPosition(),
			exoPlayer.getDuration());
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/progress/ExoPlayerFileProgressReader.kt
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.progress

import com.lasthopesoftware.bluewater.client.playback.exoplayer.PromisingExoPlayer
import com.lasthopesoftware.bluewater.client.playback.file.progress.ReadFileProgress
import com.lasthopesoftware.promises.extensions.toPromise
import com.namehillsoftware.handoff.promises.Promise
import org.joda.time.Duration

class ExoPlayerFileProgressReader(private val exoPlayer: PromisingExoPlayer) : ReadFileProgress, AutoCloseable {
	private var isClosed = false
	private var currentDurationPromise: Promise<Duration> = Promise.empty()
	private var fileProgress = Duration.ZERO

	@get:Synchronized
	override val progress: Promise<Duration>
		get() =
			currentDurationPromise
				.eventually({ promiseProgress() }, { promiseProgress() })
				.also { currentDurationPromise = it }

	private fun promiseProgress(): Promise<Duration> =
		exoPlayer.getPlayWhenReady().eventually { isPlaying ->
			if (isClosed || !isPlaying) fileProgress.toPromise()
			else {
				exoPlayer.getCurrentPosition().then { currentPosition ->
					if (currentPosition != fileProgress.millis)
						Duration.millis(currentPosition).also { fileProgress = it }
					else fileProgress
				}
			}
		}

	override fun close() {
		isClosed = true
	}
}
